sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -406,19 +406,21 @@ abstract class CastBase extends UnaryExpression with TimeZoneAwareExpression wit
         if (row.numFields > 0) {
           val st = fields.map(_.dataType)
           val toUTF8StringFuncs = st.map(castToString)
-          if (row.isNullAt(0)) {
+          if (fields(0).nullable && row.isNullAt(0)) {
```

Comment:
if `fields(0).nullable` is false, how can `row.isNullAt(0)` be true?

Reply:
(I have the same question)

Reply:
If user create dataframe from `spark.internalCreateDataFrame()`, the `row.isNullAt()` may be null even though the schema nullable is false.
For instance:
```scala
  val schema = StructType(Seq(
    StructField("x",
      StructType(Seq(
        StructField("y", IntegerType, true),
        StructField("z", IntegerType, false)
      )))))
  val rdd = spark.sparkContext.parallelize(Seq(InternalRow(InternalRow(1, null))))
  val df = spark.internalCreateDataFrame(rdd, schema)
  df.show
  // current master branch output
  //  +---------+
  //  |        x|
  //  +---------+
  //  |{1, null}|
  //  +---------+
```

Although the `spark.internalCreateDataFrame()` is sql package private API, but `spark.read.json()` and `spark.read.csv()` call it without null value handled.(the example show in pr description)

Reply:
Then we need to fix the nullability. There are so many places in the Spark codebase that relies on nullability to do optimizations. It's not possible to change all of them to not trust the nullability anymore.

Can we fix `spark.read.json()` to set the nullability correctly?

Reply:
Ok, let me try.